### PR TITLE
Reapply conversion rate into Stats API + bugfixes

### PR DESCRIPTION
**Pull request by RobertJoonas**
### Changes

This PR reapplies two reverted merge commits from https://github.com/plausible/analytics/pull/3739 and https://github.com/plausible/analytics/pull/3781 which already found their way to prod for half a day.

The reason I decided to revert those was that the CSV export started failing more often with a Sentry error we've seen before, and I decided to investigate that. Also, due to the uncertainty with the conversion rate bug in general, it was a safe play to keep it out of the public API at that point.

The way we are currently adding the CR into the browser version breakdown is broken at the moment - it doesn't crash in most cases, but it doesn't return correct conversion rate either. The recent deployment just made it crash.

The fix introduced in this PR is not pretty, and in fact, although it's much less likely, there are still ways for it to fail. If that happens though, instead of a crashing, we'll be giving back both `conversion_rate` and `total_visitors` as `"N/A"`, and reporting a message to Sentry with additional debugging details.

Longer term fix to this problem involves rebuilding the database query construction logic and asking CR from ClickHouse queries directly. That's out of scope of this PR.

### Tests
- [x] Automated tests have been added

### Changelog
- [x] Entry has been added to changelog

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated

### Dark mode
- [x] This PR does not change the UI

### CODE REVIEW

File: lib/plausible/stats/breakdown.ex [hunk 1]
```diff
@@ -693,24 +693,61 @@ defmodule Plausible.Stats.Breakdown do
     # Here, we're always only interested in the first page of results
     # - the :member filter makes sure that the results always match with
     # the items in the given breakdown_results list
-    pagination = {length(items), 1}
+    page = 1
+
+    # For browser_versions we need to fetch a lot more entries than the
```
Comment (ukutaht) [inline lib/plausible/stats/breakdown.ex:698]: 🤮

Sorry, not a reflection of your work, I'm just sad about how fundamentally fucked this whole thing is. This must have been hard to debug. For now I agree we should move forward with this workaround.

File: lib/plausible/stats/breakdown.ex [hunk 2]
```diff
@@ -693,24 +693,61 @@ defmodule Plausible.Stats.Breakdown do
...
+    # pagination limit. This is because many browsers can correspond to
+    # a single version number and we need to make sure that the results
+    # without goal filter will include all combinations of browsers and
+    # their versions that were present in the `breakdown_results` array.
+    {pagination_limit, find_match_fn} =
+      if property_atom == :browser_version do
+        pagination_limit = min(elem(pagination, 0) * 10, 3_000)
+
+        find_match_fn = fn total, conversion ->
+          total[:browser_version] == conversion[:browser_version] &&
```
Comment (ukutaht) [inline lib/plausible/stats/breakdown.ex:708]: Do we need a special case for OS versions, too?

File: lib/plausible/stats/breakdown.ex [hunk 3]
```diff
@@ -693,24 +693,61 @@ defmodule Plausible.Stats.Breakdown do
...
-        Enum.find(res_without_goal, fn s ->
-          Map.fetch!(s, property_atom) == Map.fetch!(item, property_atom)
-        end)
+      without_goal = Enum.find(res_without_goal, &find_match_fn.(&1, item))
+
+      {conversion_rate, total_visitors} =
+        if without_goal do
+          {Util.calculate_cr(without_goal.visitors, item.visitors), without_goal.visitors}
+        else
+          Sentry.capture_message(
```
Comment (ukutaht) [inline lib/plausible/stats/breakdown.ex:731]: Nice!

**Review by ukutaht — COMMENTED** [review]
Not approving atm because I'm unsure whether OS versions need the same treatment?

> Longer term fix to this problem involves rebuilding the database query construction logic and asking CR from ClickHouse queries directly. That's out of scope of this PR.

This makes sense to me 👍 I've thought about this before and it should be possible to do it all in one query since Clickouse is scanning the whole column anyways. Something like:

```sql
SELECT 
  countIf(name = 'Signup') AS conversions, 
  count(*) AS visitors ,
  conversions / visitors AS conversion_rate
FROM 
  events_v2 AS e
GROUP BY
  e.browser, e.browser_version
```

Not sure if you were thinking along the same lines. This would require substantial rewriting of core query building logic. But I think it's worth it because it solves pagination issues PLUS it allows in the future to allow ordering their reports by conversion_rate which is a must-have (see https://feedback.plausible.io/206). Also avoids an unnecessary extra query.

File: lib/plausible/stats/breakdown.ex [hunk 4]
```diff
@@ -693,24 +693,61 @@ defmodule Plausible.Stats.Breakdown do
...
+            "Unable to find a conversion_rate divisor from a breakdown response",
+            extra: %{
+              domain: site.domain,
+              query: query,
+              property: property,
+              pagination: pagination,
+              item_not_found: item
```
Comment (zoldar) [inline lib/plausible/stats/breakdown.ex:738]: I think it's better to wrap each of those passed values in `inspect(...)` as some might blow up when attempting to convert them to string (tuples for instance). The inspect form also preserves more information, usually.

File: lib/plausible/stats/breakdown.ex (hunk 1 above)
Comment (RobertJoonas) [reply lib/plausible/stats/breakdown.ex:698]: Totally. There was a lot of frustration involved in building this 😀... Can't wait to get rid of `add_cr/5` 🤞

File: lib/plausible/stats/breakdown.ex (hunk 2 above)
Comment (RobertJoonas) [reply lib/plausible/stats/breakdown.ex:708]: OS versions are not treated the same way. Yet... If you look at the [do_group_by function clause for the `os_version` property](https://github.com/plausible/analytics/blob/23b9032148b69ae27544ebd43223a6fbfdf9441e/lib/plausible/stats/breakdown.ex#L588), unlike two function clauses below for browser versions, there's no os in the group by clause.

Currently, only versions (e.g. 14, 11, 20.04) are returned from the breakdown, and many OS's can fall under the same entry. So in this PR, there's no special handling required for OS versions.

However, I have another [PR](https://github.com/plausible/analytics/pull/3806) open for adding os in the os version breakdown which will add a special case in the `add_cr` function too.

**Review by zoldar — APPROVED** [review]

**Pull request merged**